src/model.jl

```diff
@@ -102,64 +102,64 @@ _fieldparentbasetype(T, ::Type{Val{N}}) where N = T.name.wrapper
 
 # It may seem expensive always calling `param`, but flattening the
 # object occurs once at compile-time, and should have very little cost here.
-Base.length(m::AbstractModel) = length(params(m))
-Base.size(m::AbstractModel) = (length(params(m)),)
-Base.first(m::AbstractModel) = first(params(m))
-Base.last(m::AbstractModel) = last(params(m))
+Base.length(m::AbstractModel; kw...) = length(params(m; kw...))
+Base.size(m::AbstractModel; kw...) = (length(params(m; kw...)),)
+Base.first(m::AbstractModel; kw...) = first(params(m; kw...))
+Base.last(m::AbstractModel; kw...) = last(params(m; kw...))
 Base.firstindex(m::AbstractModel) = 1
-Base.lastindex(m::AbstractModel) = length(params(m))
+Base.lastindex(m::AbstractModel; kw...) = length(params(m; kw...))
 Base.getindex(m::AbstractModel, i) = getindex(params(m), i)
 Base.iterate(m::AbstractModel) = (first(params(m)), 1)
 Base.iterate(m::AbstractModel, s) = s > length(m) ? nothing : (params(m)[s], s + 1)
 
 # Vector methods
-Base.collect(m::AbstractModel) = collect(m.val)
-Base.vec(m::AbstractModel) = collect(m)
-Base.Array(m::AbstractModel) = vec(m)
+Base.collect(m::AbstractModel; kw...) = collect(m[:val; kw...])
+Base.vec(m::AbstractModel; kw...) = collect(m; kw...)
+Base.Array(m::AbstractModel; kw...) = vec(m; kw...)
 
 # Dict methods - data as columns
 Base.haskey(m::AbstractModel, key::Symbol) = key in keys(m)
 Base.keys(m::AbstractModel) = _keys(params(m), m)
 
-@inline function Base.setindex!(m::AbstractModel, x, nm::Symbol)
+@inline function Base.setindex!(m::AbstractModel, x, nm::Symbol; kw...)
     if nm == :component
         erorr("cannot set :component index")
     elseif nm == :fieldname
         erorr("cannot set :fieldname index")
     else
         newparent = if nm in keys(m)
-            _setindex(parent(m), Tuple(x), nm)
+            _setindex(parent(m), Tuple(x), nm; kw...)
         else                                 
-            _addindex(parent(m), Tuple(x), nm)
+            _addindex(parent(m), Tuple(x), nm; kw...)
         end
         setparent!(m, newparent)
     end
 end
 # TODO do this with lenses
-@inline function _setindex(obj, xs::Tuple, nm::Symbol)
+@inline function _setindex(obj, xs::Tuple, nm::Symbol; select=SELECT, ignore=IGNORE)
```

Comment:
Should `select` and `ignore` maybe have type bounds, `Type{T} where {T<:AbstractParam}`?